firebase-firestore/src/test/java/com/google/firebase/firestore/LoadBundleTaskTest.java

```diff
@@ -115,6 +123,64 @@ public void testProgressListener() throws InterruptedException {
     latch.await();
   }
 
+  @Test
+  public void testProgressListenerFiresInOrder() throws InterruptedException {
+    BlockingQueue<Integer> blockingQueue = new ArrayBlockingQueue<>(2);
+
+    LoadBundleTask task = new LoadBundleTask();
+    task.addOnProgressListener(testExecutor, progress -> blockingQueue.add(1));
+    task.addOnProgressListener(testExecutor, progress -> blockingQueue.add(2));
+
+    task.updateProgress(SUCCESS_RESULT);
+
+    assertEquals(1, (long) blockingQueue.take());
+    assertEquals(2, (long) blockingQueue.take());
+  }
+
+  @Test
+  public void testProgressListenerFireOnSpecifiedExecutor() throws InterruptedException {
+    CountDownLatch latch = new CountDownLatch(2);
+
+    LoadBundleTask task = new LoadBundleTask();
+    task.addOnProgressListener(
+        p -> {
+          assertNotEquals(TEST_THREAD_NAME, Thread.currentThread().getName());
```

Comment:
Can you confirm that this assertion failure will actually cause a test failure? Since the exception will be thrown on some other thread I just don't know if it will cause the test to fail.

Reply:
The test shows the stacktrace, but doesn't fail. I added a super ugly hack to ensure the test fails when this happens. Let me know if you can think of something better.

Reply:
Are you able to use `org.junit.rules.ErrorCollector`? If so, I think it would be a perfect fit for reporting assertion failures from other threads.

https://junit.org/junit4/javadoc/latest/org/junit/rules/ErrorCollector.html

Reply:
I saw this when I googled around, but it didn't quite strike my mind that this is exactly what I need. Thanks!